File: stacks/dynamic-dashboard/genui-controlled-shiny/scripts/prove.py
"""Evidence for the README. Costs tokens: every row is a real model call.

  consistency   — each intent prompt N times; which tool, which props (validated PlotSpec;
                  "raw nulls" counts runs where the model spelled out null for optional fields)
  refusal       — out-of-catalog prompts must produce text and zero tool calls
  cross-stack   — this stack's majority spec per intent vs the nextjs stack's Run-2 spec

Same prompts as genui-controlled-nextjs/scripts/prompts.json (vendored in fixtures/).
"""

from __future__ import annotations

import json
import sys
from collections import Counter
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from app.agent import MODEL, turn  # noqa: E402

FIX = ROOT / "fixtures"
N = int(sys.argv[1]) if len(sys.argv) > 1 else 5


def stable(v: Any) -> str:
    return json.dumps(v, sort_keys=True, separators=(",", ":"))


def main() -> int:
    prompts = json.loads((FIX / "prompts.json").read_text())
    run2 = json.loads((FIX / "nextjs-run2-specs.json").read_text())
    failures = 0

    print(f"# prove — {N} runs per intent, model {MODEL}, chatlas, {__import__('datetime').date.today()}\n")
    print("## Consistency\n")
    print("| intent | tool (n/N) | distinct specs | raw nulls | median ms | spec (validated) |")
    print("|---|---|---|---|---|---|")
    majority: dict[str, str] = {}
    for it in prompts["intents"]:
        turns = [turn(it["prompt"]) for _ in range(N)]
        names = Counter("+".join(n for n, _ in t.tools) or "(none)" for t in turns)
        specs = Counter(stable([a for _, a in t.specs]) for t in turns)
        nulls = sum(t.raw_nulls for t in turns)
        top_name, top_n = names.most_common(1)[0]
        top_spec = specs.most_common(1)[0][0]
        majority[it["id"]] = top_spec
        if len(specs) > 1:
            failures += 1
        ms = sorted(t.ms for t in turns)[N // 2]
        print(f"| {it['id']} | {top_name} ({top_n}/{N}) | {len(specs)} | {nulls}/{N} | {ms} | `{top_spec}` |")
        if len(specs) > 1:
            for s, c in specs.most_common():
                print(f"|   ↳ variant ×{c} | | | | | `{s}` |")
    print()

    print("## Refusal\n")
    print("| prompt | tool calls | reply |")
    print("|---|---|---|")
    leaks = 0
    for r in prompts["refusals"]:
        t = turn(r["prompt"])
        leaks += len(t.tools) > 0
        print(f"| {r['prompt']} | {len(t.tools)} | {t.text.replace('|', '\\|')[:220]} |")
    print(f"\nOut-of-catalog prompts that produced a plot: {leaks}/{len(prompts['refusals'])}\n")

    print("## Cross-stack — majority spec here vs the nextjs stack's Run-2 spec\n")
    print("| intent | equal | nextjs Run-2 spec |")
    print("|---|---|---|")
    mismatches = 0
    for it, expected in zip(prompts["intents"], run2, strict=True):
        exp = stable([{k: v for k, v in expected.items() if k != "kind"}])
        equal = majority[it["id"]] == exp
        mismatches += not equal
        print(f"| {it['id']} | {'✓' if equal else '✗'} | `{exp}` |")
    print(f"\nIntents whose majority spec differs from the nextjs stack: {mismatches}/{len(run2)}")
    return 1 if (failures or leaks or mismatches) else 0


if __name__ == "__main__":
    sys.exit(main())

File: stacks/dynamic-dashboard/genui-controlled-nextjs/scripts/prove.ts
/**
 * Evidence for the README, produced against the *running* app (`just dev` first).
 *
 *   consistency  — each intent prompt N times through /api/chat; which tool, which props.
 *   refusal      — out-of-catalog prompts must produce text and zero tool calls.
 *
 * Goes through the real HTTP route via DefaultChatTransport, so this measures the app,
 * not a direct model call. Prints markdown; paste it into README → Evidence.
 */
import { DefaultChatTransport, readUIMessageStream, type UIMessage } from "ai";
import { readFileSync } from "node:fs";
import path from "node:path";

const API = process.env.PROVE_API ?? "http://localhost:3000/api/chat";
const N = Number(process.env.PROVE_N ?? 5);
const prompts = JSON.parse(readFileSync(path.join(import.meta.dirname, "prompts.json"), "utf8")) as {
  intents: { id: string; prompt: string }[];
  refusals: { id: string; prompt: string }[];
};

type ToolPart = { type: string; state: string; input?: unknown; errorText?: string };
type Turn = { tools: { name: string; input: unknown }[]; text: string; ms: number };

async function turn(prompt: string): Promise<Turn> {
  const transport = new DefaultChatTransport<UIMessage>({ api: API });
  const t0 = performance.now();
  const stream = await transport.sendMessages({
    chatId: `prove-${Date.now()}`,
    messageId: undefined,
    trigger: "submit-message",
    messages: [{ id: "u1", role: "user", parts: [{ type: "text", text: prompt }] }],
    abortSignal: undefined,
  });
  let last: UIMessage | undefined;
  for await (const m of readUIMessageStream({ stream })) last = m;
  const ms = Math.round(performance.now() - t0);
  const parts = (last?.parts ?? []) as ToolPart[];
  const tools = parts
    .filter((p) => p.type.startsWith("tool-"))
    .map((p) => ({ name: p.type.slice(5), input: p.input, state: p.state, errorText: p.errorText }));
  const text = parts
    .filter((p) => p.type === "text")
    .map((p) => (p as unknown as { text: string }).text)
    .join(" ")
    .trim();
  return { tools, text, ms };
}

/** Free-text props the model may write. Drift here is copy drift, reported but not a failure. */
const COPY_FIELDS = ["title"];

function stripCopy(input: unknown): unknown {
  if (!input || typeof input !== "object") return input;
  return Object.fromEntries(Object.entries(input as Record<string, unknown>).filter(([k]) => !COPY_FIELDS.includes(k)));
}

function stable<T>(v: T): string {
  return JSON.stringify(v, (_, x) => (x && typeof x === "object" && !Array.isArray(x) ? Object.fromEntries(Object.entries(x).sort()) : x));
}

async function main() {
  console.log(`# prove — ${N} runs per intent, model ${process.env.OPENAI_MODEL ?? "(server's OPENAI_MODEL)"}, ${new Date().toISOString().slice(0, 10)}\n`);

  console.log("## Consistency\n");
  console.log("| intent | tool (n/N) | distinct specs | distinct ignoring copy fields | median ms | spec |");
  console.log("|---|---|---|---|---|---|");
  let drifts = 0;
  let copyDrifts = 0;
  for (const it of prompts.intents) {
    const runs: Turn[] = [];
    for (let i = 0; i < N; i++) runs.push(await turn(it.prompt));
    const toolNames = runs.map((r) => r.tools.map((t) => t.name).join("+") || "(none)");
    const majority = mode(toolNames);
    const specs = runs.map((r) => stable(r.tools.map((t) => t.input)));
    const structural = runs.map((r) => stable(r.tools.map((t) => stripCopy(t.input))));
    const distinct = new Set(specs);
    const distinctStructural = new Set(structural);
    if (distinctStructural.size > 1) drifts++;
    else if (distinct.size > 1) copyDrifts++;
    const ms = runs.map((r) => r.ms).sort((a, b) => a - b)[Math.floor(N / 2)];
    console.log(`| ${it.id} | ${majority} (${toolNames.filter((t) => t === majority).length}/${N}) | ${distinct.size} | ${distinctStructural.size} | ${ms} | \`${mode(specs)}\` |`);
    if (distinct.size > 1) {
      for (const [s, c] of count(specs)) console.log(`|   ↳ variant ×${c} | | | | | \`${s}\` |`);
    }
  }
  console.log(`\nIntents with structural drift (any field but ${COPY_FIELDS.join(", ")}): ${drifts}/${prompts.intents.length}`);
  console.log(`Intents with copy-only drift (${COPY_FIELDS.join(", ")}): ${copyDrifts}/${prompts.intents.length}\n`);

  console.log("## Refusal\n");
  console.log("| prompt | tool calls | reply |");
  console.log("|---|---|---|");
  let leaks = 0;
  for (const r of prompts.refusals) {
    const t = await turn(r.prompt);
    if (t.tools.length > 0) leaks++;
    console.log(`| ${r.prompt} | ${t.tools.length} | ${t.text.replace(/\|/g, "\\|").slice(0, 220)} |`);
  }
  console.log(`\nOut-of-catalog prompts that produced a plot: ${leaks}/${prompts.refusals.length}`);
  process.exit(drifts === 0 && leaks === 0 ? 0 : 1);
}

function count(xs: string[]): Map<string, number> {
  const m = new Map<string, number>();
  for (const x of xs) m.set(x, (m.get(x) ?? 0) + 1);
  return m;
}
function mode(xs: string[]): string {
  return [...count(xs).entries()].sort((a, b) => b[1] - a[1])[0]?.[0] ?? "";
}

main().catch((e) => {
  console.error(e);
  process.exit(2);
});
